src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs

```diff
@@ -174,26 +174,101 @@ void IComponent.Init(RenderHandle renderHandle)
-                OnInit();
+                return RunInitAndSetParameters();
+            }
+            else
+            {
+                OnParametersSet();
+                // If you override OnInitAsync or OnParametersSetAsync and return a noncompleted task,
+                // then by default we automatically re-render once each of those tasks completes.
+                var isAsync = false;
+                Task parametersTask = null;
+                (isAsync, parametersTask) = ProcessLifeCycletask(OnParametersSetAsync());
+                StateHasChanged();
+                // We call StateHasChanged here so that we render after OnParametersSet and after the
+                // synchronous part of OnParametersSetAsync has run, and in case there is async work
+                // we trigger another render.
+                if (isAsync)
+                {
+                    return parametersTask;
+                }
 
-                // If you override OnInitAsync and return a noncompleted task, then by default
-                // we automatically re-render once that task completes.
-                var initTask = OnInitAsync();
-                ContinueAfterLifecycleTask(initTask);
+                return Task.CompletedTask;
             }
+        }
 
-            OnParametersSet();
-            var parametersTask = OnParametersSetAsync();
-            ContinueAfterLifecycleTask(parametersTask);
+        private async Task RunInitAndSetParameters()
+        {
+            _hasCalledInit = true;
+            var initIsAsync = false;
 
+            OnInit();
+            Task initTask = null;
+            (initIsAsync, initTask) = ProcessLifeCycletask(OnInitAsync());
+            if (initIsAsync)
+            {
+                // Call state has changed here so that we render after the sync part of OnInitAsync has run
+                // and wait for it to finish before we continue. If no async work has been done yet, we want
+                // to defer calling StateHasChanged up until the first bit of async code happens or until
+                // the end.
+                StateHasChanged();
+                await initTask;
+            }
+
+            OnParametersSet();
+            Task parametersTask = null;
+            var setParametersIsAsync = false;
+            (setParametersIsAsync, parametersTask) = ProcessLifeCycletask(OnParametersSetAsync());
+            // We always call StateHasChanged here as we want to trigger a rerender after OnParametersSet and
+            // the synchronous part of OnParametersSetAsync has run, triggering another re-render in case there
+            // is additional async work.
             StateHasChanged();
+            if (setParametersIsAsync)
+            {
+                await parametersTask;
+            }
+        }
+
+        private (bool isAsync, Task asyncTask) ProcessLifeCycletask(Task task)
+        {
+            switch (task == null ? TaskStatus.RanToCompletion : task.Status)
```

Comment:
What's the use case for this to be null? Do we allow a user to return a null task from `OnInitAsync`?

Reply:
This was code that was already there. I just reshuffled it and didn't want to change it.
@SteveSandersonMS Is it important that we keep the null check?

Reply:
Historically we have tolerated lifecycle methods returning `null`. I agree we could drop that tolerance if we wanted. I'm fine either way.

Reply:
Yeah, it's a pretty weird to allow returning a null task.

Reply:
Ok, let's remove it then